process/txcost/transactionCostProcessor.go

```diff
@@ -0,0 +1,175 @@
+		if currentRes.Data.RetMessage == "" {
+			continue
+		}
+
+		res.RetMessage = currentRes.Data.RetMessage
+		res.TxCost = 0
+		return res, nil
+	}
+
+	numRes := len(tcp.responses)
+	totalGas := tcp.responses[numRes-1].Data.TxCost + initialGasLimit - tcp.responses[numRes-2].Data.TxCost
+	res.TxCost = totalGas
+
+	return res, nil
+}
+
+func (tcp *transactionCostProcessor) doCostRequests(tx *data.Transaction) (*data.TxCostResponseData, error) {
+	senderShardID, receiverShardID, err := tcp.computeSenderAndReceiverShardID(tx.Sender, tx.Receiver)
+	if err != nil {
+		return nil, err
+	}
+
+	observers, err := tcp.proc.GetObservers(receiverShardID)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, observer := range observers {
+		txCostResponse := &data.ResponseTxCost{}
+		respCode, errCall := tcp.proc.CallPostRestEndPoint(observer.Address, TransactionCostPath, tx, txCostResponse)
+		if respCode == http.StatusOK && errCall == nil {
+			return tcp.processResponse(senderShardID, receiverShardID, txCostResponse, tx)
+		}
+
+		// if observer was down (or didn't respond in time), skip to the next one
+		if respCode == http.StatusNotFound || respCode == http.StatusRequestTimeout {
+			log.LogIfError(err)
+			continue
+		}
+
+		// if the request was bad, return the error message
+		return nil, err
+
+	}
+
+	return nil, ErrSendingRequest
+}
+
+func (tcp *transactionCostProcessor) processResponse(
+	senderShardID uint32,
+	receiverShardID uint32,
+	response *data.ResponseTxCost,
+	originalTx *data.Transaction,
+) (*data.TxCostResponseData, error) {
+	tcp.responses = append(tcp.responses, response)
+
+	if len(response.Data.ScResults) == 0 || response.Data.RetMessage != "" {
+		return &response.Data, nil
+	}
+
+	for scrHash, scr := range response.Data.ScResults {
+		if scr.Used {
+			continue
+		}
+
+		scr.Used = true
+		res, err := tcp.processScResult(senderShardID, receiverShardID, scr, originalTx)
+		if err != nil {
+			log.Warn("cannot process smart contract result", "hash", scrHash, "error", err)
+			continue
+		}
+
+		if res == nil {
+			continue
+		}
+
+		mergeResponses(response, res)
+	}
+
+	return &response.Data, nil
+}
+
+func mergeResponses(finalRes *data.ResponseTxCost, currentRes *data.TxCostResponseData) {
+	for scrHash, scr := range currentRes.ScResults {
+		finalRes.Data.ScResults[scrHash] = scr
+		finalRes.Data.ScResults[scrHash].Used = true
+	}
+}
+
+func (tcp *transactionCostProcessor) processScResult(
+	senderShardID uint32,
+	receiverShardID uint32,
+	scr *data.ApiSmartContractResultExtended,
+	originalTx *data.Transaction,
+) (*data.TxCostResponseData, error) {
+	scrSenderShardID, scrReceiverShardID, err := tcp.computeSenderAndReceiverShardID(scr.SndAddr, scr.RcvAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO check if this condition is enough
+	shouldIgnoreSCR := receiverShardID == scrReceiverShardID
+	shouldIgnoreSCR = shouldIgnoreSCR || (scrReceiverShardID == senderShardID && scr.CallType == 0)
```

Comment:
use `vmcommon.DirectCall` instead of `0`

Reply:
done.